src/python/pants/backend/python/subsystems/pex_build_util.py

```diff
@@ -140,40 +144,57 @@ def subsystem_dependencies(cls):
         PythonSetup,
       )
 
+    @property
+    def generate_ptex(self) -> bool:
+      return cast(bool, self.get_options().generate_ptex)
+
     @classmethod
-    def create(cls, builder, log=None):
+    def create(cls, builder, log=None, generate_ptex=None, pex_requirement=None):
       options = cls.global_instance().get_options()
       setuptools_requirement = f'setuptools=={options.setuptools_version}'
+      pex_requirement = f'pex=={pex_requirement}' if pex_requirement else 'pex'
 
       log = log or logging.getLogger(__name__)
 
       return PexBuilderWrapper(builder=builder,
                                python_repos_subsystem=PythonRepos.global_instance(),
                                python_setup_subsystem=PythonSetup.global_instance(),
                                setuptools_requirement=PythonRequirement(setuptools_requirement),
-                               log=log)
+                               pex_requirement=PythonRequirement(pex_requirement),
+                               log=log,
+                               generate_ptex=bool(generate_ptex))
 
   def __init__(self,
-               builder,
-               python_repos_subsystem,
-               python_setup_subsystem,
-               setuptools_requirement,
-               log):
-    assert isinstance(builder, PEXBuilder)
-    assert isinstance(python_repos_subsystem, PythonRepos)
-    assert isinstance(python_setup_subsystem, PythonSetup)
-    assert isinstance(setuptools_requirement, PythonRequirement)
+               builder: PEXBuilder,
+               python_repos_subsystem: PythonRepos,
+               python_setup_subsystem: PythonSetup,
+               setuptools_requirement: PythonRequirement,
+               pex_requirement: PythonRequirement,
+               log: Any,
+               generate_ptex: bool = False):
     assert log is not None
 
     self._builder = builder
     self._python_repos_subsystem = python_repos_subsystem
     self._python_setup_subsystem = python_setup_subsystem
     self._setuptools_requirement = setuptools_requirement
+    self._pex_requirement = pex_requirement
     self._log = log
 
-    self._distributions = {}
+    self._distributions: Dict[str, Distribution] = {}
     self._frozen = False
 
+    self._generate_ptex = generate_ptex
+    # If we generate a dehydrated "ptex" file, we need to make sure that it is aware of any special
+    # find_links repos attached to any single requirement, so it can later resolve those
+    # requirements when it is first bootstrapped, using the same resolve options.
+    self._all_find_links: Set[str] = set()
```

Comment:
You want both `indexes` and `find_links` and you probably want to preserve order. You really don't know if pex / pip has an order dependency and if they do, you don't want to get in the way.